### PR TITLE
MDEV-16026: Global system_versioning_asof must not be used if client …

### PR DESCRIPTION
**Pull request by FooBarrior**
…sessions can have non-default time zone

* store `system_versioning_asof` in unix time;
* session var processed in session timezone; global var processed in global timezone;

As a regression, we cannot assign values below 1970 anymore

### CODE REVIEW

**Comment by midenok** [comment]
Please, cleanup the patch.

File: sql/sql_select.cc [hunk 1]
```diff
@@ -675,14 +675,22 @@ setup_without_group(THD *thd, Ref_ptr_array ref_pointer_array,
...
+    vars = &global_system_variables;
+
+  vers_asof_timestamp_t &in= vars->vers_asof_timestamp;
   type= (vers_system_time_t) in.type;
   start.unit= VERS_TIMESTAMP;
   if (type != SYSTEM_TIME_UNSPECIFIED && type != SYSTEM_TIME_ALL)
   {
     DBUG_ASSERT(type == SYSTEM_TIME_AS_OF);
+    MYSQL_TIME ltime;
+    thd->variables.time_zone->gmt_sec_to_TIME(&ltime, in.timestamp);
```
Comment (kevgs) [inline sql/sql_select.cc]: Global variable and session timezone. Is that ok?

File: sql/sql_select.cc [hunk 2]
```diff
@@ -675,14 +675,22 @@ setup_without_group(THD *thd, Ref_ptr_array ref_pointer_array,
 
 bool vers_select_conds_t::init_from_sysvar(THD *thd)
 {
-  vers_asof_timestamp_t &in= thd->variables.vers_asof_timestamp;
+  system_variables *vars = &thd->variables;
+  if (vars->vers_asof_timestamp.type == SYSTEM_TIME_UNSPECIFIED)
+    vars = &global_system_variables;
```
Comment (midenok) [inline sql/sql_select.cc]: That's not the way `global_system_variables` are used. Session variable must be used **always**. Global variable is used **only** when new session is created. I.e. session variable is initialized from global variable.

**Review by midenok — CHANGES_REQUESTED** [review]

**Comment by midenok** [comment]
2 things still not fixed:
1. commit message;
2. word typo (interfer).

**Comment by FooBarrior** [comment]
done

**Review by midenok — APPROVED** [review]
Use "Squash and merge" to merge to trunk.

**Pull request merged**